### PR TITLE
fix: Parent disabled property can not pass to the children when showS…

### PR DESCRIPTION
**Pull request by Francoshum95**
resolve: https://github.com/ant-design/ant-design/issues/37247

### CODE REVIEW

File: src/Cascader.tsx [hunk 1]
```diff
@@ -162,6 +162,19 @@ function toRawValues(value: ValueType): SingleValueType[] {
...
+  let newDisabled = disabled;
+
+  if (disabled || option.disabled) {
+    option.disabled = true;
+    newDisabled = true;
+  }
+
+  if (option.children) {
+    option.children.forEach(appendDisabled(newDisabled));
+  }
```
Comment (MadCcc) [inline src/Cascader.tsx:175]:
```suggestion
  option.disabled =  disabled || option.disabled;
  
  if (option.children) {
    option.children.forEach(appendDisabled(option.disabled));
  }
```

File: src/Cascader.tsx [hunk 2]
```diff
@@ -233,7 +246,17 @@ const Cascader = React.forwardRef<CascaderRef, InternalCascaderProps>((props, re
   );
 
   // =========================== Option ===========================
-  const mergedOptions = React.useMemo(() => options || [], [options]);
+
+  // Pass the parent disabled property to the children
+  const initialOptions = (newOptions: DefaultOptionType[] = []) => {
+    newOptions.forEach(newOption => {
+      if (newOption.children) {
+        newOption.children.forEach(appendDisabled(newOption.disabled));
```
Comment (MadCcc) [inline src/Cascader.tsx:254]: Use `map` to make sure `option` is not modified.

Comment (Francoshum95) [reply src/Cascader.tsx:254]:
```suggestion
  const initialOptions = (newOptions: DefaultOptionType[] = [], disabled: boolean = false) => {
    return newOptions.map(option => {
      if (disabled || option.disabled) {
        option.disabled = true;
      }

      if (option.children) {
        initialOptions(option.children, option.disabled);
      }

      return option;
    });
  };
```

How about refactoring this code? What do you think ?

File: src/Cascader.tsx [hunk 3]
```diff
@@ -233,7 +233,21 @@ const Cascader = React.forwardRef<CascaderRef, InternalCascaderProps>((props, re
   );
 
   // =========================== Option ===========================
-  const mergedOptions = React.useMemo(() => options || [], [options]);
+
+  // Pass the parent disabled property to the children
+  const initialOptions = (newOptions: DefaultOptionType[] = [], disabled: boolean = false) => {
+    return newOptions.map(option => {
+      option.disabled = disabled || option.disabled;
```
Comment (MadCcc) [inline src/Cascader.tsx:240]: Origin `option` will also be modified here.

Comment (Francoshum95) [reply src/Cascader.tsx:240]: Hi @MadCcc , sorry for my late reply


I found a way to clone each object and modify it. 
It work except one test case failed in `loadData.spec.tsx` - `'nest load should not crash'`. 

Do you mind taking a minute to review the code and kindly give some feedback

The code as follow:

```
  const initialOptions = (newOptions: DefaultOptionType[] = [], disabled: boolean = false) => {
    return newOptions.map(option => {
      const cloneOption = {...option};
      cloneOption.disabled = disabled || cloneOption.disabled;

      if (cloneOption.children) {
        cloneOption.children = initialOptions(cloneOption.children, cloneOption.disabled);
      }

      return cloneOption;
    });
  };
```

Thanks

**Pull request closed without merging**